Make the timeout decorator raise an exception out of the function's scope

abadger opened the pull request:
##### SUMMARY
Make the timeout decorator properly raise an exception outside of the function's scope
    
signal handlers which raise exceptions will never work well because the
exception can be raised anywhere in the called code.  This leads to
exception race conditions where the exceptions could end up being
hanlded by unintended pieces of the called code.
    
The timeout decorator was using just that idiom.  It was especially bad
because the decorator syntactically occurs outside of the called code
but because of the signal handler, the exception was being raised inside
of the called code.
    
This change uses a thread instead of a signal to manage the timeout in
parallel to the execution of the decorated function.  Since raising of
the exception happens inside of the decorator, now, instead of inside of
a signal handler, the timeout exception is raised from outside of the
called code as expected which makes reasoning about where exceptions are
to be expected intuitive again.
    
Fixes #43884
Fixes #49824
Fixes #49817

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request


##### COMPONENT NAME
lib/ansible/module_utils/facts/timeout.py

abadger committed:
Revert "allow caller to deal with timeout (#49449)"

This reverts commit 63279823a77320b8cdf7f69d61af6eddaa5ebf33.

Flawed on many levels

* Adds poor API to a public function
* Papers over the fact that the public function is doing something bad
  by catching exceptions it cannot handle in the first place
* Papers over the real cause of the issue which is a bug in the timeout
  decorator
* Doesn't reraise properly
* Catches the wrong exception

Fixes #49824
Fixes #49817

abadger committed:
One way to fix problems with the timeout decorator

This relies on the convention that BaseException should not be caught.
If BaseException is caught it will also catch other code that shouldn't
be caught (for instance, our fail_json and exit_json functions) however,
it is *just* convention.

If we use this fix, it would be a good idea to also try to get
45389 merged so that we're linting for bare exceptions.

Includes tests for the various ways timeouts break

abadger committed:
Make the timeout decorator properly raise an exception outside of the function's scope

signal handlers which raise exceptions will never work well because the
exception can be raised anywhere in the called code.  This leads to
exception race conditions where the exceptions could end up being
hanlded by unintended pieces of the called code.

The timeout decorator was using just that idiom.  It was especially bad
because the decorator syntactically occurs outside of the called code
but because of the signal handler, the exception was being raised inside
of the called code.

This change uses a thread instead of a signal to manage the timeout in
parallel to the execution of the decorated function.  Since raising of
the exception happens inside of the decorator, now, instead of inside of
a signal handler, the timeout exception is raised from outside of the
called code as expected which makes reasoning about where exceptions are
to be expected intuitive again.

Fixes #43884

bcoca commented on lib/ansible/module_utils/facts/timeout.py:
```diff
@@ -25,29 +26,30 @@
+        pass
+except NameError:
+    # On Python 2, there unfortunately isn't a better base class to inherit from
+    class TimeoutError(Exception):
```
in py3 TimeoutError inherits from OSError, making this code subject to divergent paths in py2 vs py3, so if you really want to avoid that inherit from OSError, not Exception

abadger replied:
Will do

bcoca commented on lib/ansible/module_utils/facts/timeout.py:
```diff
@@ -25,29 +26,30 @@
-            finally:
-                signal.alarm(0)
-            return result
+                return res.get(timeout_value)
```
I was testing this in my threading fact gathering branch,  the .get method won't help with 'blocked' processes and will inconsistently return or just hang forever (I'm still trying to get a reliable reproducer). I found this by 'simulating' bad NFS mounts by throttling the traffic to the nfs server the VMs are using.

bcoca replied:
In py3 this is mitigated by the subproccess methods having a timeout, but this is not available on py2

bcoca replied:
see my loop over waiting on threads here #49398 to get a py2/py3 compatible way

abadger replied:
We'll need a reproducer to be able to figure that out.  There's at least two other ways I can code this using threads but unless I can reproduce this I won't be able to work out what code is causing issues.

I tested this by running it on:
```
@timeout.timeout(1)
def slow_function():
    def run_command(['python', '-c', ' import time ; time.sleep(100)'])

try:
    slow_function()
except timeout.TimeoutError as e:
   print('Success: %s' % to_native(e))
```

and didn't see any problems so I'll need more to go on than that.

bcoca replied:
its not a question of slow, but 'blocking' i.e accessing a bad NFS mount, that puts the process in 'B' status for the kernel and does not allow interruption.

abadger replied:
I'm unable to reproduce with https://gist.github.com/abadger/41ddf282503b94e84512dbbb7e9f93e0 but that's uninterruptible sleep state which might be different than blocking state.  If you can write a driver.sh for blocking state we can test that instead.

bcoca replied:
Seems close enough, so I'll take that test as good as the 'unkillable' status is the main issue once the problem surfaces.

I still have not been able to reproduce the problem i saw reliably,  it just has happens a few time across the many times I've tested the gathering threaded code. It seems to happen less with Py3 versions, but since I'm not sure about how to trigger it, that is just anecdotal data. I was thinking of downgrading to nfsv3 since that was a lot more prone to this kind of issue ... but its probably not worth it.

bcoca commented:
Also, this does not fix #43884, which expects the timeout during fact gathering to be non fatal, that is what i was laying groundwork in #49449 (incorrectly since i didn't realize TimeoutError was overloaded)  as a way to allow the caller to decide if the error is fatal or not and requires a subsequent update of fact gathering to capture, note and continue with processing.

abadger commented:
This does fix #43884. @alikins' analysis of why that bug was happening was spot on here: https://github.com/ansible/ansible/issues/43884#issuecomment-412585523

The code already has an exception handler around the call to the decorated function which is expected to make the timeout non-fatal.  The problem is that to Python, the exception is being thrown at an indeterminate place in the Python code path when signals are used. To Python, the exception appears to be thrown immediately after whatever the last Python instruction processed was.  That means that when subprocess.Popen().wait() within run_command() is the last thing executed before this code is run, then python sees the exception as having originated from there.  The exception handler inside of run_command() then catches and processes the exception before the exception handler around the decorated function gets a chance to handle it.

This code removes the problem by removing the indeterminism.  Instead of the exception appearing to originate inside of the decorated function to Python, the exception orignates from the decorator, *outside of the decorated function*.  That means that no code within the decorated function will see the timeout exception.  This gives the try: except outside of the decorated function the chance to process the exception, which is exactly what the programmer would expect since they are creating the timeout decorator outside the function and they expect the exception to be generated from having added the decorator.

abadger committed:
Fix PEP8 problems and add a common case test.

Adding an integration test driven from our unittests.  Most of the time
we'll timeout in run_command which is running things in a subprocess.
Create a test for that specific case in case anything funky comes up
between threading and execve.

abadger committed:
Make TimeoutError inherit from OSError on Python2

On Python3, TimeoutError inherits from OSError so this is appropriate.

bcoca commented:
But that also makes system timeouts have the same severity, which I don't think they should

abadger committed:
Don't use OSError-based TimeoutError as a base class

Unlike most standard exceptions, OSError has a specific parameter list
with specific meanings.  Instead follow the example of other stdlib
functions, concurrent.futures and multiprocessing and define a separate
TimeoutException.

abadger commented:
You can still differentiate quite easily between the two exceptions.  And you can also use them together as would be the common desire of most code (which doesn't care what code is raising a timeout, merely that the code in question timed out).  However, I remembered that OSError is actually one of the few Python builtin exceptions which assign specific meaning to any parameters passed to constructor and that didn't seem like the proper thing to have our exception do so I researched what other stdlib libraries and saw that they simply create their own TimeoutError so I've changed the code back to that.

bcoca commented:
At least put a comment in the class, the whole reason my previous commit broke CI was because of the name collision and it appearing as a 'core' TimeoutError vs custom one (and i was testing on 3.6 which made it 'pass').

abadger committed:
Add comment and docstring to point out that this is not hte Python3 TimeoutError

bobross419 commented:
Will this be pushed to 2.6 as well?
